### PR TITLE
instancetype: Extend `PreferSpread`

### PR DESCRIPTION
**Pull request by lyarwood**
/cc @0xFelix 
/cc @fabiand 
/cc @vladikr 

**What this PR does / why we need it**:

The `preferSpread` `preferredCPUTopology` was introduced with KubeVirt v1.1 and allows for vCPUs provided by an instance type to be spread across guest visible sockets and cores using a configurable ratio that defaults to 2.

This design proposal aims to extend this implementation by introducing new configurables within a preference to control how vCPUs are spread with the goal of allowing vCPUs to also be spread to threads.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

This supersedes the original [Allow custom guest CPU topologies within instancetypes](https://github.com/kubevirt/community/pull/283) proposal given insight from a few downstream teams around their need for SMT within the guest OS for their workloads. While this proposal doesn't guarantee performant SMT aligned threads within the guest it at least allows us to express that threads will be present through the existing instance type and preference API paradigms. 

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+
+## Goals
+
+* Allow for vCPUs to be spread across all attributes of the guest visible CPU topology
+
+## User Stories
+
+* As an admin or VM owner I would like to provide preferences that spread vCPUs provided by an instance type over the guest visible sockets, cores and threads
+
+* As a VM owner I would like to consume preferences that that spread vCPUs provided by an instance type over the guest visible sockets, cores and threads
```
Comment (0xFelix) [inline design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:76]: Add note to these that mentions making use of host SMT?

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+# Design
+
+### `spec.cpu.spreadOptions`
+
+A new `spreadOptions` struct will be introduced to hold new configurables detailing how vCPUs should be spread over the guest visible CPU topology.
+
+```go
+type spreadAcross string
+
+type SpreadOptions struct {
```
Comment (0xFelix) [inline design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:93]: Do we need to make the fields pointers?

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+This option will spread vCPUs between sockets and cores. Given the current behaviour of `PreferSpread` this will remain the default.
+
+```go
+const SpreadAcrossSocketsCores SpreadAcross = "SocketsCores"
+```
+
+### `spreadAcross` `CoresThreads`
+
+This option will spread vCPUs between cores and threads with a max of 2 threads per core.
```
Comment (0xFelix) [inline design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:122]: This means ratio will be ignored?

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+
+```go
+const spreadAcrossSocketsCoresThreads SpreadAcross = "SocketsCoresThreads"
+```
+
+### `spec.preferSpreadSocketToCoreRatio` deprecation
+
+This original configurable for the `preferSpread` `preferredCPUTopology` was added directly to the core preference `spec`.
+
+This was a mistake and given the new nested `spec.cpu.spreadOptions` struct we can now deprecate the original field and move users over to `spec.cpu.spreadOptions.ratio`.
```
Comment (0xFelix) [inline design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:140]: Deprecation in `v1beta2`, removal in `v1`?

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+  "cores": 4,
+  "sockets": 2,
+  "threads": 2
+}
+```
+
+## Functional Testing Approach
+
+* Extend existing unit test coverage to ensure the expected topologies are generated within the VirtualMachineInstance.
+* Unclear if any additional functional test coverage is required outside of this.
```
Comment (0xFelix) [inline design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:321]: Tests / verification that deny SocketsThreads without dedicated CPU placement?

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md [hunk 6]
```diff
@@ -0,0 +1,325 @@
+# Overview
+
+The `preferSpread` `preferredCPUTopology` was introduced with KubeVirt v1.1 and allows for vCPUs provided by an instance type to be spread across guest visible sockets and cores using a configurable ratio that defaults to 2.
```
Comment (fabiand) [inline design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:3]: is preferSpread a field name or value?
Same for preferredCPUTopology?

Maybe reword/reformat to make it clearer, i.e.:

> The `preferredCPUTopology: preferSpread` was introduced …

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+  "sockets": 2,
+  "threads": 1
+}
+```
+
+This design proposal aims to extend this implementation by introducing new configurables within a preference to control how vCPUs are spread with the goal of allowing vCPUs to also be spread to threads.
+
+## Motivation
+
+The current implementation of the `preferSpread` `preferredCPUTopology` is limited to spreading vCPUs across sockets and cores of the guest CPU topology. While a useful starting point it would be beneficial to also allow these vCPUs to be spread across all parts of the guest visible CPU topology including threads in order to mimic SMT within the guest.
```
Comment (fabiand) [inline design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:64]: FWIW why is there the redundancy of the `prefer` string between the key and value?

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+
+This is particularly useful for performance related workloads such as DPDK etc where we would like to make use of SMT on the host in order to map through sibling threads from pCPUs to guest thread siblings. Note that this use case requires `dedicatedCPUPlacement` and `guestMappingPassthrough` to be useful so the CPUManager can allocate complete pCPUs (via the `full-pcpus-only` option) and for these pCPU siblings to be mapped correctly as vCPU thread siblings to the guest.
```
Comment (fabiand) [inline design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:66]: Given this new feature of ths priposal, are we then able to craft an instanceType which is suited for DPDK?

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+
+* kubevirt/kubevirt
+
+* kubevirt/common-instancetypes
+
+# Design
+
+### `spec.cpu.spreadOptions`
+
+A new `spreadOptions` struct will be introduced to hold new configurables detailing how vCPUs should be spread over the guest visible CPU topology.
```
Comment (fabiand) [inline design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:88]: In which API? I suspect in the VMI?

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md (hunk 2 above)
Comment (lyarwood) [reply design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:93]: Yeah we can, there's no defined best practice here as strings and ints work with `omitempty` but using pointers does make the logic for detecting if optional values are present much cleaner.

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md (hunk 3 above)
Comment (lyarwood) [reply design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:122]: I think it's more that only a ratio of 2 is supported and will need to be enforced by the validation webhook.

I'll add a note.

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md (hunk 4 above)
Comment (lyarwood) [reply design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:140]: I don't think we need a new version to start deprecation of fields?

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md (hunk 5 above)
Comment (lyarwood) [reply design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:321]: See https://github.com/kubevirt/kubevirt/issues/11749 - I wouldn't enforce that here.

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md (hunk 6 above)
Comment (lyarwood) [reply design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:3]: `preferredCPUTopology` is the field and also associated string type but ACK I'll reword this.

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md (hunk 7 above)
Comment (lyarwood) [reply design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:64]: See later in the proposal, we should clean this up and drop the `prefer` so this becomes just `Spread`.

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md (hunk 8 above)
Comment (lyarwood) [reply design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:66]: The aim is to actually provide preferences targeting DPDK and reuse the existing `CX` class of instance types from common-instancetypes.

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md (hunk 9 above)
Comment (lyarwood) [reply design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:88]: Apologies this is within the preference spec, I'll make that clear.

File: design-proposals/instancetype.kubevirt.io/extend-preferSpread.md (hunk 4 above)
Comment (0xFelix) [reply design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:140]: No, also how many steps do you still have in mind before making it v1? :)

Comment (lyarwood) [reply design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:140]: I'd hope by KubeVirt v1.4 but it all depends on feedback and closing out the remain large gaps like CR upgrades and move to `.Status` etc.

Comment (0xFelix) [reply design-proposals/instancetype.kubevirt.io/extend-preferSpread.md:140]: ACK

**Review by 0xFelix — APPROVED** [review]
Looks good to me, thank!

/lgtm

**Comment by lyarwood** [comment]
@vladikr would you mind approving this as the implementation has already landed?
